"unexpected file/folder encountered during data folder upgrade"

**Build scan**: https://gradle-enterprise.elastic.co/s/hpznznp3ugdz6/console-log?task=:qa:ccs-rolling-upgrade-remote-cluster:v7.15.0%23oneThirdUpgraded

**Repro line**: None given

**Reproduces locally?**: Not tried

**Applicable branches**: Only seen on `master`

**Failure history**: Doesn't seem to be searchable.

**Failure excerpt**:

The test fails reporting a dead node:

```
process was found dead while waiting for ports files, node{:qa:ccs-rolling-upgrade-remote-cluster:v7.15.0-remote-0}
```

The details are in the log, indicating a leftover `.es_temp_file.final` file.

```
» [2021-07-05T11:43:46,009][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [v7.15.0-remote-0] uncaught exception in thread [main] |  
-- | --
  | »  org.elasticsearch.bootstrap.StartupException: java.lang.IllegalStateException: unexpected file/folder encountered during data folder upgrade: /dev/shm/elastic+elasticsearch+master+intake+multijob+bwc/qa/ccs-rolling-upgrade-remote-cluster/build/testclusters/v7.15.0-remote-0/data/nodes/0/.es_temp_file.final |  
  | »  	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:163) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:150) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:75) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:116) ~[elasticsearch-cli-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.cli.Command.main(Command.java:79) ~[elasticsearch-cli-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:115) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:81) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  Caused by: java.lang.IllegalStateException: unexpected file/folder encountered during data folder upgrade: /dev/shm/elastic+elasticsearch+master+intake+multijob+bwc/qa/ccs-rolling-upgrade-remote-cluster/build/testclusters/v7.15.0-remote-0/data/nodes/0/.es_temp_file.final |  
  | »  	at org.elasticsearch.env.NodeEnvironment.upgradeLegacyNodeFolders(NodeEnvironment.java:386) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:268) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.node.Node.<init>(Node.java:337) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.node.Node.<init>(Node.java:261) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:226) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:226) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:392) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:159) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	... 6 more

```

--- a/server/src/main/java/org/elasticsearch/env/NodeEnvironment.java
+++ b/server/src/main/java/org/elasticsearch/env/NodeEnvironment.java
@@ -365,6 +365,13 @@ public final class NodeEnvironment  implements Closeable {
                 SNAPSHOT_CACHE_FOLDER
             ));
 
+            final Set<String> ignoredFileNames = new HashSet<>(Arrays.asList(
+                NODE_LOCK_FILENAME,
+                TEMP_FILE_NAME,
+                TEMP_FILE_NAME + ".tmp",
+                TEMP_FILE_NAME + ".final"
+            ));
+
             try (DirectoryStream<Path> stream = Files.newDirectoryStream(legacyNodePath.path)) {
                 for (Path subFolderPath : stream) {
                     final String fileName = subFolderPath.getFileName().toString();
@@ -381,8 +388,7 @@ public final class NodeEnvironment  implements Closeable {
                                 targetSubFolderPath);
                         }
                         folderNames.add(fileName);
-                    } else if (fileName.equals(NODE_LOCK_FILENAME) == false &&
-                               fileName.equals(TEMP_FILE_NAME) == false) {
+                    } else if (ignoredFileNames.contains(fileName) == false) {
                         throw new IllegalStateException("unexpected file/folder encountered during data folder upgrade: " +
                             subFolderPath);
                     }

